sql/expression/function/find_in_set.go

```diff
@@ -0,0 +1,139 @@
+
+package function
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/dolthub/go-mysql-server/sql"
+	"github.com/dolthub/go-mysql-server/sql/expression"
+	"github.com/dolthub/go-mysql-server/sql/types"
+)
+
+// FindInSet takes out the specified unit(s) from the time expression.
+type FindInSet struct {
+	expression.BinaryExpression
+}
+
+var _ sql.FunctionExpression = (*FindInSet)(nil)
+var _ sql.CollationCoercible = (*FindInSet)(nil)
+
+// NewFindInSet creates a new FindInSet expression.
+func NewFindInSet(e1, e2 sql.Expression) sql.Expression {
+	return &FindInSet{
+		expression.BinaryExpression{
+			Left:  e1,
+			Right: e2,
+		},
+	}
+}
+
+// FunctionName implements sql.FunctionExpression
+func (f *FindInSet) FunctionName() string {
+	return "find_in_set"
+}
+
+// Description implements sql.FunctionExpression
+func (f *FindInSet) Description() string {
+	return "returns a value in the range of 1 to N if the string str is in the string list strlist consisting of N substrings"
+}
+
+// Type implements the Expression interface.
+func (f *FindInSet) Type() sql.Type { return types.Int64 }
+
+// CollationCoercibility implements the interface sql.CollationCoercible.
+func (*FindInSet) CollationCoercibility(ctx *sql.Context) (collation sql.CollationID, coercibility byte) {
+	return ctx.GetCollation(), 5
+}
+
+func (f *FindInSet) String() string {
+	return fmt.Sprintf("%s(%s from %s)", f.FunctionName(), f.Left, f.Right)
+}
+
+// WithChildren implements the Expression interface.
+func (f *FindInSet) WithChildren(children ...sql.Expression) (sql.Expression, error) {
+	if len(children) != 2 {
+		return nil, sql.ErrInvalidChildrenNumber.New(f, len(children), 2)
+	}
+	return NewFindInSet(children[0], children[1]), nil
+}
+
+// Eval implements the Expression interface.
+func (f *FindInSet) Eval(ctx *sql.Context, row sql.Row) (interface{}, error) {
+	if f.Left == nil || f.Right == nil {
+		return nil, nil
+	}
+
+	left, err := f.Left.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	right, err := f.Right.Eval(ctx, row)
+	if err != nil {
+		return nil, err
+	}
+
+	if left == nil || right == nil {
+		return nil, nil
+	}
+
+	lVal, _, err := types.LongText.Convert(left)
+	if err != nil {
+		return nil, err
+	}
+	l := lVal.(string)
+
+	// always returns 0 when left contains a comma
+	if strings.Contains(l, ",") {
+		return 0, nil
+	}
+
+	var r string
+	rType := f.Right.Type()
+	if setType, ok := rType.(types.SetType); ok {
+		// TODO: set type should take advantage of bit arithmetic
+		r, err = setType.BitsToString(right.(uint64))
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		var rVal interface{}
+		rVal, _, err = types.LongText.Convert(right)
+		if err != nil {
+			return nil, err
+		}
+		r = rVal.(string)
+	}
+
+	collationPreference, _ := f.CollationCoercibility(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	strType := types.CreateLongText(collationPreference)
+	for i, r := range strings.Split(r, ",") {
+		cmp, err := strType.Compare(l, r)
```

Comment:
@Hydrocharged – is there something else we need to do to get collation aware comparisons? It seems like the string type would encapsulate that if we have the right collation preference set here, but James is still seeing the case insensitive collation test cases not match like we are expecting them to.

Reply:
Hey Jason, I talked to James about this. His implementation of `CollationCoercibility` doesn't take into account the left and right expressions, so I gave him some pointers on how to resolve it.